File: pyana_examples/trunk/src/dump_acqiris.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id$
#
# Description:
#  Pyana user analysis module dump_princeton...
#
#------------------------------------------------------------------------

"""Example module for accessing SharedIpimb data.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id$

@author Andy Salnikov
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision$"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from pypdsdata import xtc

#----------------------------------
# Local non-exported definitions --
#----------------------------------

# local definitions usually start with _

#---------------------
#  Class definition --
#---------------------
class dump_acqiris (object) :
    """Class whose instance will be used as a user analysis module. """

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, source="" ) :
        """Class constructor takes the name of the data source.

        @param source   data source
        """
        
        self.m_src = source

    #-------------------
    #  Public methods --
    #-------------------
    def beginjob( self, evt, env ) :
        try:
            env.assert_psana()
            self.psana = True
        except:
            self.psana = False
        
        config = env.getConfig(xtc.TypeId.Type.Id_AcqConfig, self.m_src)
        if config:
        
            print "%s: %s" % (config.__class__.__name__, self.m_src)
            
            print "  nbrBanks =", config.nbrBanks(),
            print "channelMask =", config.channelMask(),
            print "nbrChannels =", config.nbrChannels(),
            print "nbrConvertersPerChannel =", config.nbrConvertersPerChannel()
     
            h = config.horiz()
            print "  horiz: sampInterval =", h.sampInterval(),
            print "delayTime =", h.delayTime(),
            print "nbrSegments =", h.nbrSegments(),
            print "nbrSamples =", h.nbrSamples()

            nch = config.nbrChannels()
            for ch in range(nch):
                if self.psana:
                    v = config.vert()[ch]
                else:
                    v = config.vert(ch)
                print "  vert(%d):" % ch,
                print "fullScale =", v.fullScale(),
                print "slope =", v.slope(),
                print "offset =", v.offset(),
                print "coupling =", v.coupling(),
                print "bandwidth=", v.bandwidth()


    def event( self, evt, env ) :
        """This method is called for every L1Accept transition.

        @param evt    event data object
        @param env    environment object
        """

        if self.psana:
            acqData = evt.get("Psana::Acqiris::DataDesc", self.m_src).data_list()
        else:
            acqData = evt.get(xtc.TypeId.Type.Id_AcqWaveform, self.m_src)

        for chan, elem in enumerate(acqData):

            print "%s: %s: channel = %d" % (elem.__class__.__name__, self.m_src, chan)
            print "  nbrSegments =", elem.nbrSegments()
            print "  nbrSamplesInSeg =", elem.nbrSamplesInSeg()
            if self.psana:
                print "  timestamps =", [[elem.timestamp()[seg].pos(), elem.timestamp()[seg].timeStampHi()] for seg in range(elem.nbrSegments())]
            else:
                print "  timestamps =", [elem.timestamp(seg) for seg in range(elem.nbrSegments())]
            if self.psana:
                wf = elem.waveforms()[0] / 655360.0
            else:
                wf = elem.waveform()
            print "  waveform [len=%d] = %s" % (len(wf), wf)


    def endjob( self, evt, env ) :
        
        pass

File: psana_examples/tags/V00-04-21/src/DumpAcqiris.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpAcqiris...
//
// Author List:
//      Andrei Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpAcqiris.h"

//-----------------
// C/C++ Headers --
//-----------------
#include <algorithm>

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "pdsdata/xtc/DetInfo.hh"
#include "psddl_psana/acqiris.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpAcqiris)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpAcqiris::DumpAcqiris (const std::string& name)
  : Module(name)
{
  m_source = configSrc("source", "DetInfo(:Acqiris.*)");
}

//--------------
// Destructor --
//--------------
DumpAcqiris::~DumpAcqiris ()
{
}

// Method which is called at the beginning of the calibration cycle
void 
DumpAcqiris::beginCalibCycle(Event& evt, Env& env)
{
  MsgLog(name(), trace, "in beginCalibCycle()");

  shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(m_source, &m_src);
  if (acqConfig) {
    WithMsgLog(name(), info, str) {
      str << "Acqiris::ConfigV1: nbrBanks=" << acqConfig->nbrBanks()
          << " channelMask=" << acqConfig->channelMask()
          << " nbrChannels=" << acqConfig->nbrChannels()
          << " nbrConvertersPerChannel=" << acqConfig->nbrConvertersPerChannel();
     
      const Psana::Acqiris::HorizV1& h = acqConfig->horiz();
      str << "\n  horiz: sampInterval=" << h.sampInterval()
           << " delayTime=" << h.delayTime()
           << " nbrSegments=" << h.nbrSegments()
           << " nbrSamples=" << h.nbrSamples();
      
      const ndarray<const Psana::Acqiris::VertV1, 1>& vert = acqConfig->vert();
      for (unsigned ch = 0; ch < acqConfig->nbrChannels(); ++ ch) {
        const Psana::Acqiris::VertV1& v = vert[ch];
        str << "\n  vert(" << ch << "):"
            << " fullScale=" << v.fullScale()
            << " slope=" << v.slope()
            << " offset=" << v.offset()
            << " coupling=" << v.coupling()
            << " bandwidth=" << v.bandwidth();
      }
    }
  }
}

// Method which is called with event data
void 
DumpAcqiris::event(Event& evt, Env& env)
{

  Pds::Src src;
  shared_ptr<Psana::Acqiris::DataDescV1> acqData = evt.get(m_src);
  if (acqData) {
    
    // find matching config object
    shared_ptr<Psana::Acqiris::ConfigV1> acqConfig = env.configStore().get(m_src);
    
    // loop over channels
    int nchan = acqData->data_shape()[0];
    for (int chan = 0; chan < nchan; ++ chan) {
      
      const Psana::Acqiris::DataDescV1Elem& elem = acqData->data(chan);

      const Psana::Acqiris::VertV1& v = acqConfig->vert()[chan];
      double slope = v.slope();
      double offset = v.offset();

      WithMsgLog(name(), info, str ) {

        str << "Acqiris::DataDescV1: channel=" << chan
           << "\n  nbrSegments=" << elem.nbrSegments()
           << "\n  nbrSamplesInSeg=" << elem.nbrSamplesInSeg()
           << "\n  indexFirstPoint=" << elem.indexFirstPoint();
        
        const ndarray<const Psana::Acqiris::TimestampV1, 1>& timestamps = elem.timestamp();
        const ndarray<const int16_t, 2>& waveforms = elem.waveforms();

        // loop over segments
        for (unsigned seg = 0; seg < elem.nbrSegments(); ++ seg) {

          unsigned size = elem.nbrSamplesInSeg();
          ndarray<const int16_t, 1> raw(waveforms[seg]);
          ndarray<float, 1> wf = make_ndarray<float>(size);
          for (unsigned i = 0; i < size; ++ i) {
            wf[i] = raw[i]*slope + offset;
          }
          
          str << "\n  Segment #" << seg
              << "\n    timestamp value = " << timestamps[seg].value() << " pos = " << timestamps[seg].pos()
              << "\n    raw = " << raw
              << "\n    data = " << wf;
        
        }
      }
    }
  }
}

} // namespace psana_examples
